deployment/common/changeset/state_test.go

```diff
@@ -0,0 +1,203 @@
+	}
+	tests := []testCase{
+		{
+			name:  "Valid contracts",
+			chain: defaultChain,
+			addresses: map[string]deployment.TypeAndVersion{
+				"0x123": deployment.NewTypeAndVersion(types.RBACTimelock, deployment.Version1_0_0),
+				"0x456": deployment.NewTypeAndVersion(types.CallProxy, deployment.Version1_0_0),
+				"0x789": deployment.NewTypeAndVersion(types.ProposerManyChainMultisig, deployment.Version1_0_0),
+				"0xabc": deployment.NewTypeAndVersion(types.CancellerManyChainMultisig, deployment.Version1_0_0),
+				"0xdef": deployment.NewTypeAndVersion(types.BypasserManyChainMultisig, deployment.Version1_0_0),
+			},
+			wantState: &MCMSWithTimelockState{
+				MCMSWithTimelockContracts: &proposalutils.MCMSWithTimelockContracts{
+					Timelock: func() *owner_helpers.RBACTimelock {
+						tl, err := owner_helpers.NewRBACTimelock(common.HexToAddress("0x123"), nil)
+						require.NoError(t, err)
+						return tl
+					}(),
+					CallProxy: func() *owner_helpers.CallProxy {
+						cp, err := owner_helpers.NewCallProxy(common.HexToAddress("0x456"), nil)
+						require.NoError(t, err)
+						return cp
+					}(),
+					ProposerMcm: func() *owner_helpers.ManyChainMultiSig {
+						mcms, err := owner_helpers.NewManyChainMultiSig(common.HexToAddress("0x789"), nil)
+						require.NoError(t, err)
+						return mcms
+					}(),
+					CancellerMcm: func() *owner_helpers.ManyChainMultiSig {
+						mcms, err := owner_helpers.NewManyChainMultiSig(common.HexToAddress("0xabc"), nil)
+						require.NoError(t, err)
+						return mcms
+					}(),
+					BypasserMcm: func() *owner_helpers.ManyChainMultiSig {
+						mcms, err := owner_helpers.NewManyChainMultiSig(common.HexToAddress("0xdef"), nil)
+						require.NoError(t, err)
+						return mcms
+					}(),
+				},
+			},
+			wantErr: "",
+		},
+		{
+			name:  "Missing contract",
+			chain: defaultChain,
+			addresses: map[string]deployment.TypeAndVersion{
+				"0x123": deployment.NewTypeAndVersion(types.RBACTimelock, deployment.Version1_0_0),
+				"0x456": deployment.NewTypeAndVersion(types.CallProxy, deployment.Version1_0_0),
+				"0x789": deployment.NewTypeAndVersion(types.ProposerManyChainMultisig, deployment.Version1_0_0),
+				"0xabc": deployment.NewTypeAndVersion(types.CancellerManyChainMultisig, deployment.Version1_0_0),
+				// "0xdef": deployment.NewTypeAndVersion(types.BypasserManyChainMultisig, deployment.Version1_0_0), // Missing
+			},
+			wantState: nil,
+			wantErr:   "missing a contract",
+		},
+		{
+			name: "labelled multichain contract proposer role overrides existing proposer",
+			// TODO(mstreet3): Skipping because this behavior is undefined.  The addresses are a map and there is no guarantee that
+			// a labeled contract will override a named contract type.  Leaving this test case in until this behavior is resolved.
+			skip:  true,
+			chain: defaultChain,
+
+			// addresses has two proposer contract candidates, which should make it into the final state?
```

Comment:
@AnieeG please advise on expected behavior. Will fix in this PR.

Reply:
Based on the switch case flow if `ProposerManyChainMultisig` is found it will just return the contract with type `ProposerManyChainMultisig` first right? That's why I have put the `ManyChainMultisig` at the end with the labels. If it does not explicitly define `ProposerManyChainMultisig` ,`CancellerManyChainMultisig` , `BypasserManyChainMultisig` only then it should go to `ManyChainMultisig`block, unless I am missing something here.

Reply:
Do you think it's better to call that out in comments or enforce that behavior apart from relying on the switch case flow?

Reply:
I totally missed it's a map.
 I think the behavior should be -
if explicitly `ProposerManyChainMultiSig` type is mentioned, it should be selected for `ProposerMcms`. Ideal way would be to only manage it with labels but I am not sure if that would break already existing migration compatibility. For now to me it makes sense to override `ManyChainMultiSig` with the label `proposer` with `ProposerManyChainMultiSig` Or we can throw an error if both are provided in addresses